File: sdk/ios/Sources/Valydar/ValydarClient.swift
import Foundation

public class ValydarClient {
    private let apiKey: String
    private let baseURL: URL
    private let session: URLSession

    public init(apiKey: String, baseURL: String = "https://api.dev.valydar.com") {
        self.apiKey = apiKey
        self.baseURL = URL(string: baseURL)!
        let config = URLSessionConfiguration.default
        config.httpAdditionalHeaders = ["Authorization": "Bearer \(apiKey)"]
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    // MARK: - Verifications

    public func createVerification(
        clientReference: String? = nil,
        checks: [String]? = nil
    ) async throws -> VerificationResponse {
        var body: [String: Any] = [
            "client_reference": clientReference ?? "ios_\(Int(Date().timeIntervalSince1970))"
        ]
        if let checks { body["checks"] = checks }
        return try await post("/verifications", body: body)
    }

    public func getVerification(id: String) async throws -> VerificationResponse {
        try await get("/verifications/\(id)")
    }

    // MARK: - Documents

    public func uploadDocument(
        verificationId: String,
        imageData: Data,
        filename: String = "document.jpg",
        documentType: String? = nil
    ) async throws -> DocumentUploadResponse {
        var components = URLComponents()
        components.scheme = baseURL.scheme
        components.host = baseURL.host
        components.path = "/verifications/\(verificationId)/documents"

        var request = URLRequest(url: components.url!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        let boundary = "Boundary-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)

        if let docType = documentType {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"document_type\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(docType)\r\n".data(using: .utf8)!)
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(DocumentUploadResponse.self, from: data)
    }

    // MARK: - Face Match

    public func faceMatch(
        verificationId: String,
        documentId: String,
        selfieId: String
    ) async throws -> FaceMatchResponse {
        try await post(
            "/verifications/\(verificationId)/face-match",
            body: ["document_id": documentId, "selfie_id": selfieId]
        )
    }

    // MARK: - Health

    public func health() async throws -> HealthResponse {
        try await get("/health")
    }

    // MARK: - Internal

    private func get<T: Decodable>(_ path: String) async throws -> T {
        let url = baseURL.appendingPathComponent(path)
        let (data, _) = try await session.data(from: url)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func post<T: Decodable>(_ path: String, body: Any) async throws -> T {
        let url = baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: sdk/android/valydar/src/main/java/com/valydar/ValydarClient.kt
package com.valydar

import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.util.concurrent.TimeUnit

class ValydarClient(
    private val apiKey: String,
    private val baseUrl: String = "https://api.dev.valydar.com"
) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true }

    private fun request(path: String): Request.Builder {
        return Request.Builder()
            .url("$baseUrl$path")
            .addHeader("Authorization", "Bearer $apiKey")
    }

    private inline fun <reified T> get(path: String): T {
        val resp = client.newCall(request(path).get().build()).execute()
        val body = resp.body?.string() ?: throw RuntimeException("Empty response")
        return json.decodeFromString(body)
    }

    private inline fun <reified T> post(path: String, body: Any? = null): T {
        val mediaType = "application/json".toMediaType()
        val jsonBody = body?.let { json.encodeToString(it) }
        val req = request(path)
            .post(jsonBody?.toRequestBody(mediaType) ?: "".toRequestBody())
            .build()
        val resp = client.newCall(req).execute()
        val result = resp.body?.string() ?: throw RuntimeException("Empty response")
        return json.decodeFromString(result)
    }

    fun health(): HealthResponse = get("/health")

    fun createVerification(
        clientReference: String = "android_${System.currentTimeMillis()}",
        checks: List<String> = listOf("document", "face_match", "liveness")
    ): VerificationResponse {
        return post("/verifications", mapOf(
            "client_reference" to clientReference,
            "checks" to checks
        ))
    }

    fun getVerification(id: String): VerificationResponse = get("/verifications/$id")

    fun uploadDocument(
        verificationId: String,
        imageFile: File,
        documentType: String? = null
    ): DocumentUploadResponse {
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("file", imageFile.name,
                imageFile.readBytes().toRequestBody("image/jpeg".toMediaType()))
            .apply { documentType?.let { addFormDataPart("document_type", it) } }
            .build()

        val req = request("/verifications/$verificationId/documents")
            .post(body)
            .build()

        val resp = client.newCall(req).execute()
        val result = resp.body?.string() ?: throw RuntimeException("Empty response")
        return json.decodeFromString(result)
    }

    fun faceMatch(
        verificationId: String,
        documentId: String,
        selfieId: String
    ): FaceMatchResponse {
        return post("/verifications/$verificationId/face-match", mapOf(
            "document_id" to documentId,
            "selfie_id" to selfieId
        ))
    }
}
